### PR TITLE
Refactor taylorize

### PR DESCRIPTION
**Pull request by lbenet**
This PR implements some ideas outlined in #155, preallocating up to 3-indices arrays of `Taylor1{T}` objects.

### CODE REVIEW

**Comment by lbenet** [comment]
Some naive benchmarks using `tests/taylorize.jl` (on second run), similar to [these](https://github.com/PerezHz/TaylorIntegration.jl/pull/152#issuecomment-1324363506), using latest Julia:

- Using this branch ([3849ecc](https://github.com/PerezHz/TaylorIntegration.jl/pull/156/commits/3849ecc782d57c262c862cee1c6ac4ed96c1ed5b)):
```julia
julia> @time include("test/taylorize.jl")
...
Test Summary:          | Pass  Total   Time
Testing `taylorize.jl` |  363    363  28.5s
 36.544097 seconds (87.81 M allocations: 8.841 GiB, 4.95% gc time, 46.77% compilation time: 2% of which was recompilation)
```

- Using master ([5422e9b](https://github.com/PerezHz/TaylorIntegration.jl/commit/5422e9b95b82662903de4c3f814b666ad5d78d9b)):
```julia
julia> @time include("test/taylorize.jl")
...
Testing `taylorize.jl` |  358    358  46.7s
 55.119991 seconds (551.97 M allocations: 23.015 GiB, 7.31% gc time, 32.01% compilation time: 2% of which was recompilation)
```

So, time and allocations seem to have been improved, with respect to master!

**Comment by lbenet** [comment]
Results from `run_benchmarks.jl`, comparing this branch with master, similar to [this comparison](https://github.com/PerezHz/TaylorIntegration.jl/pull/152#issuecomment-1327683604). I report here only the "ManySpin" benchmarks, since it was there where `@taylorize` was not useful. (The Kepler benchmarks show mixed results, some improvements and some regressions, all below 10%.)

| ID                            | time ratio                   | memory ratio                 |
|-------------------------------|------------------------------|------------------------------|
| `["ManySpin", "manyspin1-1"]` | 0.09 (5%) :white_check_mark: | 0.20 (1%) :white_check_mark: |
| `["ManySpin", "manyspin1-2"]` | 0.09 (5%) :white_check_mark: | 0.03 (1%) :white_check_mark: |
| `["ManySpin", "manyspin2-1"]` | 0.07 (5%) :white_check_mark: | 0.11 (1%) :white_check_mark: |
| `["ManySpin", "manyspin2-2"]` | 0.07 (5%) :white_check_mark: | 0.01 (1%) :white_check_mark: |

In all cases the improvements are important!

Same as the table above, but comparing with respect to `test/v0.8.12` branch ([459227](https://github.com/PerezHz/TaylorIntegration.jl/commit/459227fd8e484adfdb51ffe495e7b45e37fc8ff9))

| ID                            | time ratio                   | memory ratio                 |
|-------------------------------|------------------------------|------------------------------|
| `["ManySpin", "manyspin1-1"]` | 0.62 (5%) :white_check_mark: | 0.55 (1%) :white_check_mark: |
| `["ManySpin", "manyspin1-2"]` | 0.65 (5%) :white_check_mark: | 0.12 (1%) :white_check_mark: |
| `["ManySpin", "manyspin2-1"]` | 0.69 (5%) :white_check_mark: | 0.41 (1%) :white_check_mark: |
| `["ManySpin", "manyspin2-2"]` | 0.67 (5%) :white_check_mark: | 0.07 (1%) :white_check_mark: |

So, I think this PR reverts indeed the problems described in #155.

**Comment by lbenet** [comment]
@PerezHz @LuEdRaMo Could you check that with the version of `@taylorize` in this PR, both [PlanetaryEphemeris](https://github.com/PerezHz/PlanetaryEphemeris.jl) and [NEOs.jl](https://github.com/PerezHz/NEOs.jl) show improvements?

In the meantime I hope to improve the documentation and coverage...

**Comment by lbenet** [comment]
Motivated by [this comment](https://github.com/PerezHz/PlanetaryEphemeris.jl/issues/9#issue-1503666701) I have included support for up to 4-index arrays. What is missing is throwing an error when there are arrays with more indices...

**Comment by LuEdRaMo** [comment]
@lbenet I think 
https://github.com/PerezHz/TaylorIntegration.jl/blob/b5bc66e994f108c770570aab2fd7fde74654bf75/src/parse_eqs.jl#L1099
should include 
```julia
@inbounds for (ind, vnew) in enumerate(bkkeep.v_array4)
    push!(tmp_defs, :($(vnew) = __ralloc.v4[$(ind)]))
end
```
to declare the 4-index Taylor variables.

**Comment by lbenet** [comment]
You are right, @LuEdRaMo; the last commit follows your suggestion, and throws an error for 5-indices (or more) arrays

**Comment by lbenet** [comment]
Tests are failing in Julia v1.6 only, seemingly due to precompilation of OrdinaryDiffEq. I'll try to solve this problem; ideas are welcome.

File: test/taylorize.jl [hunk 1]
```diff
@@ -1058,8 +1093,23 @@ import Logging: Warn
...
+        ex = :(function err_arr_indx!(Dz, z, p, t)
+            n = size(z,1)
+            arr5 = Array{typeof(z[1])}(undef,n,1,1,1,1)
+            for i in eachindex(z)
+                arr5[i,1,1,1,1] = 0.0
+                Dz[i] = z[i] + arr5[i,1,1,1,1]
+            end
+            nothing
+        end)
+        @test_throws "Error: `@taylorize` allows only to parse up tp 5-index arrays" TI._make_parsed_jetcoeffs(ex)
```
Comment (PerezHz) [inline test/taylorize.jl:1111]: Maybe the error here can be fixed by changing this line with ``@test_throws ErrorException("Error: `@taylorize` allows only to parse up tp 5-index arrays") TI._make_parsed_jetcoeffs(ex)``?

**Comment by PerezHz** [comment]
> Tests are failing in Julia v1.6 only, seemingly due to precompilation of OrdinaryDiffEq. I'll try to solve this problem; ideas are welcome.

I reckon the OrdinaryDiffEq precompilation error disappeared after [cc048d](https://github.com/PerezHz/TaylorIntegration.jl/pull/156/commits/8cc048db36b706e50084a936575f25bc91307d27)?

File: test/taylorize.jl (hunk 1 above)
Comment (lbenet) [reply test/taylorize.jl:1111]: I'm checking this right now 😄; I'll push a change soon...

**Comment by lbenet** [comment]
I noticed the same locally, after cleaning up the Manifest...

**Comment by lbenet** [comment]
@PerezHz except for tagging a new version, I think this is ready for a review...

**Comment by lbenet** [comment]
Regarding the last two commits: https://github.com/PerezHz/TaylorIntegration.jl/pull/156/commits/fa2735383a4e7dd178b40555be554e9eb3e48406 is a small change so `t::Taylor1` uses always the full order of the series, when calling the function with the ODE, and https://github.com/PerezHz/TaylorIntegration.jl/pull/156/commits/272a4dacb8175f10f7bfc816fdaa4f6972cb7a0a uncomments some bechmarks. The former is motivated by an observation of @LuEdRaMo, who noticed that in certain cases, such "performance trick" induces some problems; and the second is to check that not big performance penalty was introduced.

**Comment by lbenet** [comment]
@LuEdRaMo @PerezHz I think this is ready to be merged. If you are not in favor of that, please rise your voice 😄

I'll wait until Tuesday, merge this PR and register a new minor version.

**Comment by PerezHz** [comment]
@lbenet thank you so much for this! LGTM, with "G" standing for "great"! 😄

**Comment by lbenet** [comment]
Merged! 🎉 

cc @PerezHz @LuEdRaMo

**Pull request merged**